CoreServices: Add ControlIndicatorTimer to remove GuiTick dependency

Holzhaus opened the pull request:
Previously, the `ControlIndicator` objects relied on the
`[Master],guiTickTime` and `[Master],guiTick50ms` that are managed by
the `GuiTick` class and processed in the vsync thread during waveform
processing. But the blinking of indicator controls is unrelated to
Waveforms and should still work when we get rid of the legacy skin
system and waveforms widgets.

This simple change introduces two new control objects:

- `[Master],indicator_250millis`
- `[Master],indicator_500millis`

Both switch between 0.0 and 1.0 every 250/500ms. `ControlIndicator` has
been adapted to use these new controls instead.

These controls may also be used by controller mappings that want to
implement custom blinking controls. Before, this was already possible
using a timer in JavaScript, but these were not synchronized with the
other blinking controls, such as `[ChannelN],play_indicator` or
`[ChannelN],cue_indicator` which made them more annoying than necessary.

Holzhaus committed:
CoreServices: Add ControlIndicatorTimer to remove GuiTick dependency

Previously, the `ControlIndicator` objects relied on the
`[Master],guiTickTime` and `[Master],guiTick50ms` that are managed by
the `GuiTick` class and processed in the vsync thread during waveform
processing. But the blinking of indicator controls is unrelated to
Waveforms and should still work when we get rid of the legacy skin
system and waveforms widgets.

This simple change introduces two new control objects:

- `[Master],indicator_250millis`
- `[Master],indicator_500millis`

Both switch between 0.0 and 1.0 every 250/500ms. `ControlIndicator` has
been adapted to use these new controls instead.

These controls may also be used by controller mappings that want to
implement custom blinking controls. Before, this was already possible
using a timer in JavaScript, but these were not synchronized with the
other blinking controls, such as `[ChannelN],play_indicator` or
`[ChannelN],cue_indicator` which made them more annoying than necessary.

uklotzde commented on src/control/controlindicator.cpp:
```diff
@@ -21,33 +20,39 @@ ControlIndicator::ControlIndicator(const ConfigKey& key)
+        // Disconnect old signals
         switch (m_blinkValue) {
+        case RATIO1TO1_250MS:
+            disconnect(m_pCOIndicator250millis, nullptr, this, nullptr);
```
m_pCOIndicator250millis->disconnect(this);

uklotzde commented on src/control/controlindicatortimer.cpp:
```diff
@@ -0,0 +1,32 @@
+
+void ControlIndicatorTimer::slotTimeout() {
+    // The timeout uses an interval of 250ms, so the 250ms indicator is always updated.
+    m_pCOIndicator250millis->forceSet(m_pCOIndicator250millis->toBool() ? 0.0 : 1.0);
```
Why not simply `1 - m_pCOIndicator250millis->get()`?

uklotzde commented on src/control/controlindicatortimer.cpp:
```diff
@@ -0,0 +1,32 @@
+    m_timer.start(250);
+}
+
+void ControlIndicatorTimer::slotTimeout() {
```
I suggest to use a lambda for this slot and get rid of the member variable.

Holzhaus replied:
Can you elaborate? I don't see how to avoid the member variable and use a lambda here. Except if I use two timers, which is undesired because then there is no guarantee that both indicators are in sync.

uklotzde replied:
Doesn't work. In C++ we can't capture mutable from the environment like in Rust.

uklotzde reviewed (commented):
How about using a dedicated thread? The UI event loop where this object is supposed to live causes latency. Probably over-engineering, could be refactored later.

uklotzde commented on src/control/controlindicator.h:
```diff
@@ -17,24 +22,24 @@ class ControlIndicator : public ControlObject {
   private slots:
-    void slotGuiTick50ms(double cpuTime);
     void slotBlinkValueChanged();
+    void set(double value) {
```
Overwriting a non-virtual function from the parent is confusing. Please rename as `slotValueChanged` and then simply invoke `set()`.

Holzhaus replied:
Done,

uklotzde commented on src/control/controlindicator.h:
```diff
@@ -17,24 +22,24 @@ class ControlIndicator : public ControlObject {
 
   private:
     void toggle(double duration);
     // set() is private, use setBlinkValue instead
```
This comment is now misplaced. And I also don't understand it. setBlinkValue() serves a different purpose than set().

uklotzde replied:
Just delete the comment.

uklotzde commented on src/control/controlindicator.h:
```diff
@@ -17,24 +22,24 @@ class ControlIndicator : public ControlObject {
+    };
 
   private:
     void toggle(double duration);
```
The definition of toggle() has been deleted.

daschuer commented on src/control/controlindicatortimer.cpp:
```diff
@@ -0,0 +1,32 @@
+    m_pCOIndicator250millis->setReadOnly();
+    m_pCOIndicator500millis->setReadOnly();
+    connect(&m_timer, &QTimer::timeout, this, &ControlIndicatorTimer::slotTimeout);
+    m_timer.start(250);
```
The QTimer has special code that ensures to deliver the signal as close as possible to the given time.
That sounds appealing on the fist sight, unfortunately this introduces kind of priority for the connected events that interfere with the vsync code that has a hard deadline to met to avoid jerking. If a blink event happens during vsync, it is delayed together with the whole GUI until the next buffer is swapped. If there is pending waveform work to do, it is shiftet to the next buffer which is causing a jerk.

That was the original reason to schedule the repainting of blinking widgets with the vsync thread.

This is probably not relevant for QML widgets where the vsync is probably done later, but it still applies for CPU rendered widgets. Can we make the time source conditionally depending on the skin type?

I assume that we need some kind of vsync interval for any waveform solution.
So maybe it is best to provide an generic interface for such syncing.

What do you think?

Be-ing replied:
Would running the timer in its own thread solve this?

Holzhaus replied:
> The QTimer has special code that ensures to deliver the signal as close as possible to the given time.

The default timer type is [`Qt::CoarseTimer`](https://doc.qt.io/qt-5/qt.html#TimerType-enum), that "try to keep accuracy within 5% of the desired interval". It's not a `Qt::PreciseTimer` which would try to keep millisecond accuracy.

> That sounds appealing on the fist sight, unfortunately this introduces kind of priority for the connected events that interfere with the vsync code that has a hard deadline to met to avoid jerking. If a blink event happens during vsync, it is delayed together with the whole GUI until the next buffer is swapped. If there is pending waveform work to do, it is shiftet to the next buffer which is causing a jerk.

If that was a problem, wouldn't that apply to *all* QTimers in the GUI thread? If so, why didn't we notice issues yet? We use QTimers often:

```
$ grep -Hrn "\bQTimer\b" ../src
../src/control/controlbehavior.h:3:#include <QTimer>
../src/control/controlbehavior.h:151:    QTimer* getTimer() {
../src/control/controlbehavior.h:153:            m_pushTimer.reset(new QTimer());
../src/control/controlbehavior.h:159:    QScopedPointer<QTimer> m_pushTimer;
../src/control/controlindicatortimer.h:4:#include <QTimer>
../src/control/controlindicatortimer.h:27:    QTimer m_timer;
../src/control/controlindicatortimer.cpp:17:    connect(&m_timer, &QTimer::timeout, this, &ControlIndicatorTimer::slotTimeout);
../src/widget/wtime.cpp:14:    m_pTimer = new QTimer(this);
../src/widget/wtime.cpp:25:    connect(m_pTimer, &QTimer::timeout, this, &WTime::refreshTime);
../src/widget/wsearchlineedit.h:6:#include <QTimer>
../src/widget/wsearchlineedit.h:95:    QTimer m_debouncingTimer;
../src/widget/wsearchlineedit.h:96:    QTimer m_saveTimer;
../src/widget/wtime.h:6:#include <QTimer>
../src/widget/wtime.h:25:    QTimer* m_pTimer;
../src/widget/wsearchlinee$ grep -HrnF QTimer ../src
../src/control/controlbehavior.h:3:#include <QTimer>
../src/control/controlbehavior.h:151:    QTimer* getTimer() {
../src/control/controlbehavior.h:153:            m_pushTimer.reset(new QTimer());
../src/control/controlbehavior.h:159:    QScopedPointer<QTimer> m_pushTimer;
../src/control/controlindicatortimer.h:4:#include <QTimer>
../src/control/controlindicatortimer.h:27:    QTimer m_timer;
../src/control/controlindicatortimer.cpp:17:    connect(&m_timer, &QTimer::timeout, this, &ControlIndicatorTimer::slotTimeout);
../src/vinylcontrol/vinylcontrolmanager.cpp:169:void VinylControlManager::timerEvent(QTimerEvent*) {
../src/vinylcontrol/vinylcontrolsignalwidget.h:13:#include <QTimerEvent>
../src/vinylcontrol/vinylcontrolmanager.h:5:#include <QTimerEvent>
../src/vinylcontrol/vinylcontrolmanager.h:43:    void timerEvent(QTimerEvent* pEvent);
../src/mixer/playerinfo.h:8:#include <QTimerEvent>
../src/mixer/playerinfo.h:49:    void timerEvent(QTimerEvent* pTimerEvent) override;
../src/mixer/playerinfo.cpp:112:void PlayerInfo::timerEvent(QTimerEvent* pTimerEvent) {
../src/widget/wtime.cpp:14:    m_pTimer = new QTimer(this);
../src/widget/wtime.cpp:25:    connect(m_pTimer, &QTimer::timeout, this, &WTime::refreshTime);
../src/widget/wsearchlineedit.h:6:#include <QTimer>
../src/widget/wsearchlineedit.h:95:    QTimer m_debouncingTimer;
../src/widget/wsearchlineedit.h:96:    QTimer m_saveTimer;
../src/widget/wtime.h:6:#include <QTimer>
../src/widget/wtime.h:25:    QTimer* m_pTimer;
../src/widget/wsearchlineedit.cpp:116:            &QTimer::timeout,
../src/widget/wsearchlineedit.cpp:120:            &QTimer::timeout,
../src/widget/wcoverart.h:7:#include <QTimer>
../src/widget/wcoverart.h:76:    QTimer m_clickTimer;
../src/widget/wpushbutton.h:8:#include <QTimer>
../src/widget/wpushbutton.h:110:    QTimer m_clickTimer;
../src/widget/wlibrarysidebar.cpp:110:void WLibrarySidebar::timerEvent(QTimerEvent *event) {
../src/widget/wlibrarysidebar.h:10:#include <QTimerEvent>
../src/widget/wlibrarysidebar.h:26:    void timerEvent(QTimerEvent* event) override;
../src/library/dlgcoverartfullsize.h:5:#include <QTimer>
../src/library/dlgcoverartfullsize.h:50:    QTimer m_clickTimer;
../src/library/sidebarmodel.h:6:#include <QTimer>
../src/library/sidebarmodel.h:90:    QTimer* const m_pressedUntilClickedTimer;
../src/library/sidebarmodel.cpp:34:          m_pressedUntilClickedTimer(new QTimer(this)) {
../src/library/sidebarmodel.cpp:37:            &QTimer::timeout,
../src/waveform/guitick.cpp:1:#include <QTimer>
../src/dialog/dlgdevelopertools.cpp:92:void DlgDeveloperTools::timerEvent(QTimerEvent* pEvent) {
../src/dialog/dlgdevelopertools.h:6:#include <QTimerEvent>
../src/dialog/dlgdevelopertools.h:20:    void timerEvent(QTimerEvent* pTimerEvent) override;
../src/util/timer.h:112:// A timer that provides a similar API to QTimer but uses render events from the
../src/controllers/scripting/legacy/controllerscriptinterfacelegacy.cpp:471:void ControllerScriptInterfaceLegacy::timerEvent(QTimerEvent* event) {
../src/controllers/scripting/legacy/controllerscriptinterfacelegacy.h:64:    virtual void timerEvent(QTimerEvent* event);
../src/controllers/controllermanager.h:4:#include <QTimer>
../src/controllers/controllermanager.h:77:    QTimer m_pollTimer;
../src/controllers/dlgcontrollerlearning.h:6:#include <QTimer>
../src/controllers/dlgcontrollerlearning.h:83:    QTimer m_firstMessageTimer;
../src/controllers/dlgcontrollerlearning.h:84:    QTimer m_lastMessageTimer;
../src/controllers/dlgcontrollerlearning.cpp:137:    connect(&m_lastMessageTimer, &QTimer::timeout, this, &DlgControllerLearning::slotTimerExpired);
../src/controllers/dlgcontrollerlearning.cpp:142:            &QTimer::timeout,
../src/controllers/controller.h:4:#include <QTimerEvent>
../src/controllers/controllermanager.cpp:101:    connect(&m_pollTimer, &QTimer::timeout, this, &ControllerManager::pollDevices);
../src/network/jsonwebtask.cpp:6:#include <QTimerEvent>
../src/network/webtask.cpp:4:#include <QTimerEvent>
../src/network/webtask.cpp:300:void WebTask::timerEvent(QTimerEvent* event) {
../src/network/webtask.h:146:    void timerEvent(QTimerEvent* event) final;
```dit.cpp:116:            &QTimer::timeout,
../src/widget/wsearchlineedit.cpp:120:            &QTimer::timeout,
../src/widget/wcoverart.h:7:#include <QTimer>
../src/widget/wcoverart.h:76:    QTimer m_clickTimer;
../src/widget/wpushbutton.h:8:#include <QTimer>
../src/widget/wpushbutton.h:110:    QTimer m_clickTimer;
../src/library/dlgcoverartfullsize.h:5:#include <QTimer>
../src/library/dlgcoverartfullsize.h:50:    QTimer m_clickTimer;
../src/library/sidebarmodel.h:6:#include <QTimer>
../src/library/sidebarmodel.h:90:    QTimer* const m_pressedUntilClickedTimer;
../src/library/sidebarmodel.cpp:34:          m_pressedUntilClickedTimer(new QTimer(this)) {
../src/library/sidebarmodel.cpp:37:            &QTimer::timeout,
../src/waveform/guitick.cpp:1:#include <QTimer>
../src/util/timer.h:112:// A timer that provides a similar API to QTimer but uses render events from the
../src/controllers/controllermanager.h:4:#include <QTimer>
../src/controllers/controllermanager.h:77:    QTimer m_pollTimer;
../src/controllers/dlgcontrollerlearning.h:6:#include <QTimer>
../src/controllers/dlgcontrollerlearning.h:83:    QTimer m_firstMessageTimer;
../src/controllers/dlgcontrollerlearning.h:84:    QTimer m_lastMessageTimer;
../src/controllers/dlgcontrollerlearning.cpp:137:    connect(&m_lastMessageTimer, &QTimer::timeout, this, &DlgControllerLearning::slotTimerExpired);
../src/controllers/dlgcontrollerlearning.cpp:142:            &QTimer::timeout,
../src/controllers/controllermanager.cpp:101:    connect(&m_pollTimer, &QTimer::timeout, this, &ControllerManager::pollDevices);
```

Be-ing replied:
I think accuracy within 5% should be fine for blinking these COs.

Be-ing replied:
> If a blink event happens during vsync, it is delayed together with the whole GUI until the next buffer is swapped.

This seems to be a misunderstanding as @Holzhaus explained above. Testing this branch, I do not see any regression in waveform jerking and the indicator controls continue to work.

Be-ing replied:
I switched between this branch and main several times playing 2 decks with 2 decks paused using the Pioneer cue mode which has (annoyingly) frequent blinking and did not notice a difference in waveform scrolling.

Be-ing commented on src/control/controlindicatortimer.h:
```diff
@@ -0,0 +1,33 @@
+/// This class provides two generic indicator controls that change their value
+/// in intervals of 250 milliseconds and 500 milliseconds. These controls are
+/// used by all `ControlIndicator` COs and may also be used to implement custom
+/// blinking buttons (e.g. for Explicit Sync Leader) on controllers. This
```
huh?

Holzhaus replied:
Do you have a suggestion what to change? I think it makes sense to explain what these COs are used for and that they are not "internal", i.e. that they are useful for third party code, too. I was planning to update the Roland DJ-505 to use them, because the blinking of the SUNC button not being in sync (hah! ;-) with the CUE button is annoying.

Be-ing replied:
I don't think controller scripts should be implementing custom blinking in general.

Holzhaus replied:
In general I agree that we should avoid blinking. But in some cases it can be useful. My sync button only has a single color, and in the rare cases I deem it necessary to set a sync leader explicitly, I want to know which deck is sync leader.

ronso0 replied:
active loop, playing sampler, key set off etc. all deserve blinking if the user considers that state unusual/remarkable

Be-ing commented on src/coreservices.h:
```diff
@@ -35,7 +38,7 @@ class CoreServices : public QObject {
   public:
     CoreServices(const CmdlineArgs& args);
-    ~CoreServices() = default;
+    ~CoreServices();
```
Why did you move this to the .cpp file?

Holzhaus replied:
I'm using a `std::unique_ptr` for `ControlIndicatorTimer`. In case we destruct `CoreServices` without calling `shutdown()` first (shouldn't happen, but the compiler does not know this), the compiler needs the full declaration of `ControlIndicatorTimer` to be able to destruct `ControlIndicatorTimer`.

I didn't want to include `controlindicatortimer.h` in `coreservices.h`, so I used a forward declaration. If we put the destructor definition into `coreservices.h` instead of `coreservices.cpp`, the full delcaration of `ControlIndicatorTimer` is not known and will lead to a compiler error about calling `sizeof` with an incomplete type.

Be-ing replied:
okay

Be-ing commented on src/control/controlindicatortimer.cpp:
```diff
@@ -0,0 +1,32 @@
+
+void ControlIndicatorTimer::slotTimeout() {
+    // The timeout uses an interval of 250ms, so the 250ms indicator is always updated.
+    m_pCOIndicator250millis->forceSet(1 - m_pCOIndicator250millis->get());
```
This is an odd way to toggle a boolean value.
```suggestion
    m_pCOIndicator250millis->forceSet(!m_pCOIndicator250millis->toBool());
```

uklotzde replied:
...because storing a boolean as double is odd ;) But I agree, for the sake of readability use this less efficient variant.

Holzhaus replied:
lol, we went full circle with this :D I think I prefer the original version then. That makes it clear that we're storing doubles.

```suggestion
    m_pCOIndicator250millis->forceSet(m_pCOIndicator250millis->toBool() ? 0.0 : 1.0);
```

Be-ing replied:
sure that works

Be-ing commented:
> How about using a dedicated thread? The UI event loop where this object is supposed to live causes latency. Probably over-engineering, could be refactored later.

The current implementation seems to work fine, so I think a separate thread would be overengineering.

Be-ing commented:
Thank you for working on this. This is much better than entangling CoreServices with the QWidgets GUI.

Be-ing commented:
DeckVisuals is still coupled to WaveformWidgetFactory. I suspect that can be refactored to use a similar architecture driven by a QTimer in the main thread. That can be resolved in another PR.

Be-ing commented:
Do we still need GuiTick and GuiTickTimer now? I am not sure, but I am inclined to say let's not fix what isn't broken. I don't want to reopen the can of worms that led to implementing WidgetRenderTimer.

Holzhaus commented:
> Do we still need GuiTick and GuiTickTimer now? I am not sure, but I am inclined to say let's not fix what isn't broken. I don't want to reopen the can of worms that led to implementing WidgetRenderTimer.

The guiTick COs are still used by `WidgetRenderTimer`, `WTrackTable` and the battery widget. I wasn't planning to touch that code, so let's just leave it in for legacy skins and remove GuiTick/GuiTickTimer when we remove support for these skins.

daschuer commented:
> Would running the timer in its own thread solve this? 

No, the timer source is the issue, CPU clock vs GPU clock.

> If that was a problem, wouldn't that apply to all QTimers in the GUI thread? If so, why didn't we notice issues yet? We use QTimers often:

In general we should avoid timers in the GUI thread, that triggers a repainting, not v-synced. The issue is that every timer makes the situation a tiny bit worse. You can argue case by case that it is not worth the extra effort but the sume of all will finally make the difference. 
  
> I think accuracy within 5% should be fine for blinking these COs.

The 5% is used to do similar timed actions in one callback. That is not related to the issue with v-sync, because that is independent. The current solution in main considers the v-sync interval as well. If this where also a QTimer it would work well, but it is not, unfortunately. 


> testing this branch, I do not see any regression in waveform jerking and the indicator controls continue to work.

@Be-ing It is hard to test. Please describe you test setup in details.

daschuer reviewed (changes requested):
All the performance discussion is kind of mood compared to the view lines of code that are required to use the v-sync thread as alternative timing source for these new objects. 
So I would prefer to add these lines and continue migrating to a new waveform rendering without introducing regressions into the old in the meantime. 

@Holzhaus What do you think?

Holzhaus commented:
> All the performance discussion is kind of mood compared to the view lines of code that are required to use the v-sync thread as alternative timing source for these new objects.
> So I would prefer to add these lines and continue migrating to a new waveform rendering without introducing regressions into the old in the meantime.

I honestly don't know how to implement this. I could of course add a controlproxy for the guiTick COs and use that if the CO exists, otherwise use a qtimer. But this won't work, because the guiTick CO would be leaked and stop being updated when we switch to a QML skin.

Also, are you really sure that changing the appearance of a widget depending on the CO value actually takes immediate effect and may cause issues with VSync? In my understanding, something like `QAbstractButton::setIcon()` will probably call `QWidget::update()` internally: 

> This function does not cause an immediate repaint; instead it schedules a paint event for processing when Qt returns to the main event loop. This permits Qt to optimize for more speed and less flicker than a call to repaint() does.

And if *that* conflicts with vsync, we have a much bigger issue at hand (or maybe don't, because we didn't notice any issues yet).

daschuer commented:
> I honestly don't know how to implement this.

The new object can be either connected to the vsync thread If exist or to the QTimer. Something like: if (old skin) start vsync thread; else start QTimer

daschuer commented:
> Also, are you really sure that changing the appearance of a widget depending on the CO value actually takes immediate effect and may cause issues with VSync?

My knowledge is based on extensive tests and code review of older QT versions, so I am not certain if all of this still applies, but the basic principles should still exists. 

Qt has its own vsync code that avoid tearing.
In a risk of tearing graphic updates are delated after the buffer swap. The main thread is suspended unable to do anything else then. 

Our vsync thread does the same in a different thread to avoid these extra suspends. If we sync all GUI actions with our v-sync thread, we can avoid many of these extra suspends, that leads to a single delayed waveform update. 

> In my understanding, something like QAbstractButton::setIcon() will probably call QWidget::update() internally (it just looked it up, it does):

Yes that's correct. That causes extra jerking when interact with the widgets by not synced user input. Fortunately less notable than jerking during Mixxx ideling around where the user may whatch the waveform. 

The cited doc explains how qt does it by default.
Our issue is that we have the need for our own, because the Qt way is optimized for desktop Apps not die Video Playback.

Both interfere each other. Syncing the timer with our own lints the issue, because it avoid suspends in the Qt implementation.

daschuer commented:
The best way to solve the issue for both would be to receive a v-synced timer out of the Qt anti tearing code. To find out that is on my to do list. 
Unfortunately that is buried deep inside target depending code. 
Any help is Welcome.

uklotzde commented on src/control/controlindicator.cpp:
```diff
@@ -21,33 +20,39 @@ ControlIndicator::ControlIndicator(const ConfigKey& key)
 ControlIndicator::~ControlIndicator() {
```
`ControlIndicator::~ControlIndicator() = default`

uklotzde commented:
@daschuer I suggest that we start with this simple solution unless you are able to demonstrate that it causes severe issues.

We could revert this decision at any time, no need to block progress. Let's follow YAGNI. The issues might only affect outdated Qt versions. No one knows and it has not been documented properly.

Be-ing commented:
> @Be-ing It is hard to test. Please describe you test setup in details.

I [did](https://github.com/mixxxdj/mixxx/pull/4157#discussion_r678828672). What more details do you want?

uklotzde commented:
> > @Be-ing It is hard to test. Please describe you test setup in details.
> 
> I [did](https://github.com/mixxxdj/mixxx/pull/4157#discussion_r678828672). What more details do you want?

If users or even developers are not noticing issues it is not worth fixing ;)

Holzhaus committed:
ControlIndicatorTimer: Add support for legacy vsyncthread mode

Due to concerns that a regular QTimer that fires during vsync could
cause jerking legacy waveforms, this implements a way to use the
legacy vsync thread/GuiTick infrastructure instead of a `QTimer` in the
`ControlIndicatorTimer` (although I'm still not convinced that
this actually makes a difference).

This can be toggled explicitly using `setLegacyVsyncEnabled` when
switching between a legacy and a QML skin (the latter has no vsync
thread, so a regular timer needs to be used).

Relevant discussion can be found here:
https://github.com/mixxxdj/mixxx/pull/4157#issuecomment-888945348

Be-ing commented:
> In general we should avoid timers in the GUI thread, that triggers a repainting, not v-synced. The issue is that every timer makes the situation a tiny bit worse. You can argue case by case that it is not worth the extra effort but the sume of all will finally make the difference.

The purpose of this PR is to work towards replacing the entire QWidgets GUI system so I don't think such long term concerns matter. That is assuming your understanding of the issue is correct, which I am not sure is the case, and this concern might not be a problem at all. 

> The best way to solve the issue for both would be to receive a v-synced timer out of the Qt anti tearing code. To find out that is on my to do list.

This seems like more overengineering and I suspect it is not relevant to Qt Quick. The legacy waveform rendering is working okay enough; let's not fix what isn't broken.

Be-ing commented:
@uklotzde @Holzhaus @ronso0 can you try testing [as I described](https://github.com/mixxxdj/mixxx/pull/4157#discussion_r678828672) and confirm you do not notice any regression in waveform jerking?

It would be good to confirm there is no regression on macOS too where we have had lots of mysterious performance problems. @foss- can you test the GitHub Actions artifact from this PR with the [setup I described](https://github.com/mixxxdj/mixxx/pull/4157#discussion_r678828672) and confirm that waveforms do not jerk anymore than before?

Holzhaus commented:
Whoops, I didn't notice the discussion here and already implemented the legacy vsync mode (although I still doubt that it's needed): https://github.com/mixxxdj/mixxx/pull/4157/commits/fb86a2797e7d7af2edaf42421d0952b45f242fca

We can later simply revert that commit when we drop legacy skins.

Be-ing commented on src/control/controlindicatortimer.cpp:
```diff
@@ -0,0 +1,68 @@
+    m_timer.start(250);
+}
+
+void ControlIndicatorTimer::slotTimeout() {
```
```suggestion
void ControlIndicatorTimer::slotQTimerTimeout() {
```
to be clear this is unrelated to the legacy vsync code below

Holzhaus replied:
It's still called by the legacy code below. I separated the code in both the header file and the cpp file, so I think there shouldn't be much confusion.

Be-ing replied:
okay

Be-ing commented:
> should I remove it now because of YAGNI and only apply it if someone actually notices waveform jerking?

It seems to work the same either way. It is not much code so I do not mind keeping it.

daschuer commented:
I have tested this with the VSyncTest (GL) waveform at 60 Hz and Pioneer mode enabled on Ubuntu Groovy.
A constantly pink waveform is an indicator for perfect vsync rendering.  

I can confirm that this branch now has no regression compared to main. 
I experience some rare jerks in the rhythm of the button blinking here and in main. As far as I remember there was no such dependency originally. So there must be something changes under the hood with the qt anti tearing code that is now less optimal for us. 

With the QTimer version the situation is worse. I see some kind of interference interval. A flawlessly pink wavefront for some seconds is cycling with heavy jerking also for some seconds.

Unfortunately I am not able to record this, because the recording software "SimpleScreenRecoder" is not able to record every frame vsynced. 

@Holzhaus Thank you for the legacy fix.

daschuer commented on src/control/controlindicatortimer.cpp:
```diff
@@ -0,0 +1,68 @@
+
+void ControlIndicatorTimer::slotTimeout() {
+    // The timeout uses an interval of 250ms, so the 250ms indicator is always updated.
+    m_pCOIndicator250millis->forceSet(m_pCOIndicator500millis->toBool() ? 0.0 : 1.0);
```
```suggestion
    m_pCOIndicator250millis->forceSet(m_pCOIndicator250millis->toBool() ? 0.0 : 1.0);
```

Holzhaus replied:
Good catch

daschuer commented on src/control/controlindicatortimer.cpp:
```diff
@@ -0,0 +1,68 @@
+    if (m_toggleIndicator500millisOnNextTimeout) {
+        m_pCOIndicator500millis->forceSet(m_pCOIndicator500millis->toBool() ? 0.0 : 1.0);
+    }
+    m_toggleIndicator500millisOnNextTimeout = !m_toggleIndicator500millisOnNextTimeout;
```
Idea: you can get rid of this extra bool by using a local with the result from m_pCOIndicator250millis->toBool() instead.

Holzhaus commented:
OK, ready to merge then?

daschuer reviewed (changes requested):
There is one remaining copy and paste issue to fix.

Holzhaus committed:
ControlIndicatorTimer: Remove unnecessary member variable

The `m_toggleIndicator500millisOnNextTimeout` can be removed, because
the 500ms indicator just needs to be toggled on every second toggle of
the 250ms indicator.

Be-ing commented:
@daschuer merge?

daschuer reviewed (approved):
LGTM

uklotzde commented:
Good job, everyone!
